lib/Conversion/ImportVerilog/Structure.cpp

```diff
@@ -694,6 +709,20 @@ Context::convertModuleHeader(const slang::ast::InstanceBodySymbol *module) {
       return {};
     }
   }
+
+  // Mapping the hierarchical names into the module's ports.
+  for (auto &hierPath : hierPaths[module]) {
+    auto hierType = convertType(hierPath.valueSym->getType());
+    if (!hierType)
+      return {};
+
+    // All hierarchical names are regarded as output ports with ref<T> types.
+    if (auto hierName = hierPath.hierName) {
+      hierType = moore::RefType::get(cast<moore::UnpackedType>(hierType));
+      hierPath.idx = outputIdx++;
+      modulePorts.push_back({hierName, hierType, hw::ModulePort::Output});
+    }
+  }
```

Comment:
From yesterday to now, I've always solved an error that `hierPath.idx` is empty. Due to the acquiescently deduplicate module/instance, we cannot record **subC2**. In other words, `hierPaths[subC2]` is empty, therefore, `hierPath.idx` is also empty. For example:
```
module Bar;
  SubC subC1();
  SubC subC2();
  int u = subC1.subD.z;
  assign subC2.subD.z = u;
endmodule
```

Reply:
By the way, when I use `llvm::errs()<<module->parentInstance->name` before executing deduplicate, the terminator will display `subC1` and `subC2`. If after executing deduplicate, it will display `subC1`, empty :thinking:.

Reply:
Hmmm I don't think you can make all hierarchical paths output ports. I think there are two different kinds of hierarchical references a module has to deal with:

- **Imported references.** If you are inside a module `Top.subA.subB` and the module uses the hierarchical name `Top.subA.x`, then a reference to `x` is imported as an input ref port. The parent module has to know about this and has to forward the reference to `x` to its child `subB`.
- **Exported references.** If you are inside a module `Top.subA.subB` and the parent module accesses the hierarchical name `subB.x`, then a reference to `x` is exported as an output ref port. The parent module then uses that reference for its access to `subB.x`.

Basically, for every module in the design you need a list of hierarchical references that need to be imported or exported by that module. During your initial traversal of the design where you visit all hierarchical paths, you will need to iterate through every element of a hierarchical path, look up the corresponding module, and add an entry to its list of hierarchical references. Depending on whether the hierarchical path goes downwards into a child module, or starts further up in a parent, you will have to mark the hierarchical reference as either import or export. (Children need to export references as output ports to their parents. Hierarchical paths that start further up in the hierarchy have to first export references upwards to the root of the path, and then import that references down to the hierarchical path use.)

Reply:
Really appreciate for the detailed explanation.